Implementing the ProblemConnection interface to replace S3Connection for local development

charlesnguyen8 opened the pull request:
I created and implemented the ProblemConnection interface, and created a brief design but not implemented the AssignmentConnection interface. I also modified Upload.java and Codecheck.java to use the ProblemConnection interface accordingly. 
(This pull request is to contribute to issue #23, For development, store assignments in file system)

charlesnguyen8 committed:
created 2 interfaces ProblemConnection and AssignmentConnection

charlesnguyen8 committed:
implement write() for ProblemLocalConnection

charlesnguyen8 committed:
modified Upload.java, start using ProblemConnector instead of S3Connection

charlesnguyen8 committed:
implement read() for ProblemLocalConnection

charlesnguyen8 committed:
implement the other 2 write methods and delete methods for ProblemLocalConnection

charlesnguyen8 committed:
cleaned up code, delete 2 unecessary write functions and 2 isOnS3

cayhorstmann reviewed (changes requested):
Does the injection work? I wasn't sure you can have a nested class as a @Singleton. If not, it would make sense to make ProblemConnector a top-level public class. There is no need for the ProblemConnection interface and the subclasses to be public.

Is there an advantage to nesting the classes ProblemS3Connection and ProblemLocalConnection inside ProblemConnection? If so, change their names to S3 and Local. Otherwise make them non-public classes in the file.

Same comments for AssignmentConnection. 

In AssignmentConnection,

```
public boolean isOnS3(String repo);
public boolean isOnS3(String repo, String key);
```

are not used. Drop them.

The method
```
 public ObjectNode readJsonObjectFromDB(String tableName, String primaryKeyName, String primaryKeyValue) throws IOException {
            String result = readJsonStringFromDB(tableName, primaryKeyName, primaryKeyValue);
            return result == null ? null : (ObjectNode)(new ObjectMapper().readTree(result)); 
}
```
can become a default method in the interface. That way, you don't have to implement it twice.

Same with `readJsonObjectFromDB(String tableName, String primaryKeyName, String primaryKeyValue, String sortKeyName, String sortKeyValue)`

Why do you need the methods
```
  private AmazonS3 getS3Connection() { 
      return amazonS3; 
  }

  public AmazonDynamoDB getAmazonDynamoDB() {
     return amazonDynamoDB;
  }
```
Just access the fields directly.

charlesnguyen8 committed:
update ProblemConnection structure and AssignmentConnection structure

charlesnguyen8 commented:
The injection worked even when ProblemConnector was nested class, but it does not make much sense to let it nested so I followed your advice and made it top-level class. Same thing happened to ProblemS3Connection and ProblemLocalConnection, there is no benefit to them being nested inside ProblemConnection so I moved them out and made both of them and the ProblemConnection interface non-public. 

All other suggestion such as the default methods readJsonObjectFromDB, remove isOnS3 from AssignmentConnection, accessing amazonS3 and amazonDB directly instead of through a method were also implemented.

One question I have is that if ProblemConnection need isOnS3? Currently I removed them from the interface.
